### PR TITLE
Add IP-based rules

### PR DESCRIPTION
**Pull request by Gargron**
Admins get the ability to add rules for IP addresses and CIDR ranges, such as blocking all access, or requiring sign-ups to go through the approval process even if they're otherwise open.

image

The CLI utility is extended with the following commands:

- `tootctl ip_blocks add 1.1.1.1 2.2.2.2 --severity=no_access --comment="These are bad" --duration=31557600`
- `tootctl ip_blocks remove 2.2.2.2`
- `tootctl ip_blocks export --format=nginx`

### CODE REVIEW

File: lib/mastodon/ip_blocks_cli.rb [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+      failed    = 0
+
+      addresses.each do |address|
+        ip_block = begin
+          if options[:force]
+            IpBlock.find_by('ip >>= ?', address)
+          else
+            IpBlock.find_by(ip: address)
+          end
+        end
```
Comment (ThibG) [inline lib/mastodon/ip_blocks_cli.rb:89]: This doesn't cover the case where there are multiple IP blocks covering the same address (either net blocks with some wider than others, or multiple `IpBlock` records with the exact same target, as those aren't required to be unique).

File: app/models/user.rb [hunk 2]
```diff
@@ -385,7 +387,17 @@ def render_and_send_devise_message(notification, *args)
...
+      if sign_up_from_ip_requires_approval?
+        false
+      else
+        open_registrations? || valid_invitation? || external?
+      end
+    end
+  end
+
+  def sign_up_from_ip_requires_approval?
+    !sign_up_ip.nil? && IpBlock.where(severity: :sign_up_requires_approval).where('ip >>= ?', sign_up_ip.to_s).exists?
```
Comment (ThibG) [inline app/models/user.rb:400]: That check is possibly quite slow.

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
Requesting a few changes through inline comments.

Besides that, I'm a bit worried about how much this would scale: I have to admit I have no idea how many IP/subnets an instance would typically block, but I'm a bit concerned with the code loading the full list into RAM at once, as well as enumerating all subnet blocks to find a matching one. But maybe that's fine.

**Comment by Gargron** [comment]
> Besides that, I'm a bit worried about how much this would scale: I have to admit I have no idea how many IP/subnets an instance would typically block, but I'm a bit concerned with the code loading the full list into RAM at once, as well as enumerating all subnet blocks to find a matching one. But maybe that's fine.

I have two pieces of data to offer. One, by comparing to e-mail domain blocks I've accumulated over 4 years: 64, and a list of IPs I've blocked on iptables level: 291. So the assumption I am working with is that we're dealing with three digit numbers here, which is also why I'm not quick to add indices to this table.

**Review by ClearlyClaire — APPROVED** [review]
alright, looks good to me

still a bit worried about scaling, but based on your earlier comment, it should be fine

**Comment by Kjwon15** [comment]
How about allowlist for ip rules? it could be useful for school

**Comment by rodti** [comment]
This is a great feature. A couple of nice-to-haves are to have the list sorted numerically, and to be able to edit existing IP rules (like the federation edit screen).

**Pull request merged**